dotCMS/src/main/java/com/dotcms/rest/api/v1/page/PageResourceHelper.java

```diff
@@ -141,44 +185,54 @@ public PageView getPageMetadataRendered(final HttpServletRequest request, final
     }
 
+    public String getPageRendered(final HttpServletRequest request, final HttpServletResponse response, final User user,
+                                  final String uri, PageMode mode) throws Exception {
+
+        final HTMLPageAsset page =  this.getPage(request, user, uri, mode);
+        return this.getPageRendered(page, request, response, user, mode);
+    }
+
     @CloseDB
-    public String getPageRendered(final HttpServletRequest request, final
-    HttpServletResponse response, final User user, final String uri, PageMode mode) throws DotDataException, DotSecurityException, IOException {
-        
+    public String getPageRendered(final HTMLPageAsset page, final HttpServletRequest request,
+                                  final HttpServletResponse response, final User user, PageMode mode) throws Exception {
+
         final String siteName = null == request.getParameter(Host.HOST_VELOCITY_VAR_NAME) ?
                 request.getServerName() : request.getParameter(Host.HOST_VELOCITY_VAR_NAME);
         final Host site = this.hostWebAPI.resolveHostName(siteName, user, RESPECT_FE_ROLES);
 
-        final String pageUri = (uri.length()>0 && '/' == uri.charAt(0)) ? uri : ("/" + uri);
-        final HTMLPageAsset page =  (HTMLPageAsset) this.htmlPageAssetAPI.getPageByPath(pageUri,
-                site, this.languageAPI.getDefaultLanguage().getId(), mode.respectAnonPerms);
-
         if(null==page) {
             response.sendError(HttpServletResponse.SC_NOT_FOUND);
             return null;
         }
         if(mode.isAdmin ) {
             APILocator.getPermissionAPI().checkPermission(page, PermissionLevel.READ, user);
         }
-        
+
         switch (mode) {
             case PREVIEW_MODE:
-                return new VelocityPreviewMode(request, response, pageUri, site).eval();
+                return new VelocityPreviewMode(request, response, page.getURI(), site).eval();
 
             case EDIT_MODE:
-                return new VelocityEditMode(request, response, pageUri, site).eval();
+                return new VelocityEditMode(request, response, page.getURI(), site).eval();
 
             default:
-                return new VelocityLiveMode(request, response, pageUri, site).eval();
+                return new VelocityLiveMode(request, response, page.getURI(), site).eval();
 
         }
-        
-        
-  
     }
 
-    
-    
+    public HTMLPageAsset getPage(final HttpServletRequest request, final User user,
+                                 final String uri, final PageMode mode) throws DotSecurityException, DotDataException {
+
+        final String siteName = null == request.getParameter(Host.HOST_VELOCITY_VAR_NAME) ?
+                request.getServerName() : request.getParameter(Host.HOST_VELOCITY_VAR_NAME);
+        final Host site = this.hostWebAPI.resolveHostName(siteName, user, RESPECT_FE_ROLES);
+
+        final String pageUri = (uri.length()>0 && '/' == uri.charAt(0)) ? uri : ("/" + uri);
```

Comment:
you might want to create a method here

pageUri = URLUtils.addSlashIfNeeded(uri);

The method name sucks, probably would need a better one